src/sync_state.py

```diff
@@ -0,0 +1,653 @@
+        "nickName",
+    }
+)
+
+
+def _filter_sensitive_attributes(attributes: dict[str, str]) -> dict[str, str]:
+    """Filter out sensitive attributes (names) from logging.
+
+    Args:
+        attributes: User attributes dictionary.
+
+    Returns:
+        Filtered dictionary without sensitive attributes.
+    """
+    return {k: v for k, v in attributes.items() if k not in _SENSITIVE_ATTRIBUTES}
+
+
+@dataclass(frozen=True)
+class GroupMembershipState:
+    """Current state of a group's membership.
+
+    Attributes:
+        group_id: The Identity Store group ID.
+        group_name: Human-readable group name.
+        current_members: Set of user principal IDs currently in the group.
+    """
+
+    group_id: str
+    group_name: str
+    current_members: frozenset[str]
+
+
+@dataclass(frozen=True)
+class UserInfo:
+    """Basic user information for sync operations.
+
+    Attributes:
+        user_id: The Identity Store user principal ID.
+        email: User's email address.
+        attributes: Dictionary of user attribute name to value.
+    """
+
+    user_id: str
+    email: str
+    attributes: dict[str, str]
+
+
+@dataclass(frozen=True)
+class SyncAction:
+    """An action to take during sync.
+
+    Attributes:
+        action_type: Type of action - "add", "remove", or "warn".
+        user_id: The user principal ID to act on.
+        user_email: The user's email address for logging/notifications.
+        group_id: The group ID to modify.
+        group_name: Human-readable group name for logging/notifications.
+        reason: Human-readable explanation of why this action is needed.
+        matched_attributes: Attributes that matched (for add) or were expected (for remove/warn).
+    """
+
+    action_type: Literal["add", "remove", "warn"]
+    user_id: str
+    user_email: str
+    group_id: str
+    group_name: str
+    reason: str
+    matched_attributes: dict[str, str] | None = None
+
+
+class SyncStateManager:
+    """Computes required sync actions based on current and desired state.
+
+    This class evaluates users against attribute mapping rules and compares
+    with current group memberships to determine what actions are needed.
+    """
+
+    def __init__(
+        self,
+        managed_group_ids: dict[str, str],
+        mapper: AttributeMapper,
+        manual_assignment_policy: Literal["warn", "remove"],
+    ) -> None:
+        """Initialize the sync state manager.
+
+        Args:
+            managed_group_ids: Mapping of group name to group ID for managed groups.
+            mapper: AttributeMapper instance with configured rules.
+            manual_assignment_policy: Policy for handling manual assignments.
+        """
+        self._managed_group_ids = managed_group_ids
+        self._managed_group_names = {v: k for k, v in managed_group_ids.items()}
+        self._mapper = mapper
+        self._manual_assignment_policy = manual_assignment_policy
+
+    def compute_sync_actions(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 49 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 98% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this function or break it into multiple functions.